stdlib/public/Synchronization/Mutex/LinuxImpl.swift

```diff
@@ -0,0 +1,137 @@
+//
+//===----------------------------------------------------------------------===//
+
+import SynchronizationShims
+
+@_alwaysEmitIntoClient
+@_extern(c, "gettid")
+func _gettid() -> UInt32
+
+extension Atomic where Value == UInt32 {
+  borrowing func wait() {
+    _swift_stdlib_wait(.init(rawAddress))
+  }
+
+  borrowing func wake() {
+    _swift_stdlib_wake(.init(rawAddress))
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+@frozen
+@usableFromInline
+@_staticExclusiveOnly
+internal struct _MutexHandle: ~Copyable {
+  // There are only 3 different values that storage can hold at a single time.
+  // 0: unlocked
+  // _gettid: locked, current thread's id (uncontended)
+  // (_gettid | SWIFT_FUTEX_WAITERS): locked, current thread's id (contended)
+  @usableFromInline
+  let storage: Atomic<UInt32>
+
+  @available(SwiftStdlib 5.11, *)
+  @_alwaysEmitIntoClient
+  @_transparent
+  init() {
+    storage = Atomic(0)
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func lock() {
+    // TODO: Is it worth caching this value in TLS?
+    let selfId = _gettid()
+
+    // Note: We could probably merge this cas into a do/while style loop, but we
+    // really want to perform the strong variant before attempting to do weak
+    // ones in the loop.
+
+    var (exchanged, state) = storage.compareExchange(
+      expected: 0,
+      desired: selfId,
+      successOrdering: .acquiring,
+      failureOrdering: .relaxed
+    )
+
+    if _fastPath(exchanged) {
+      // Locked!
+      return
+    }
+
+    while !exchanged {
+      // Clear the waiter bit, if we have one, and check to see if this is a
+      // recursive call to lock.
+      let currentOwner = state & ~SWIFT_FUTEX_WAITERS
+      if _slowPath(currentOwner == selfId) {
+        // TODO: Need a colder function.
+        fatalError("Recursive call to lock")
+      }
+
+      // Block until unlock has been called. This will return early if the call
+      // to unlock happened between attempting to acquire and attempting to
+      // wait while nobody else managed to acquire it yet.
+      storage.wait()
+
+      (exchanged, state) = storage.weakCompareExchange(
+        expected: 0,
+        desired: selfId,
+        successOrdering: .acquiring,
+        failureOrdering: .relaxed
+      )
+    }
+
+    // Locked!
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func tryLock() -> Bool {
+    storage.compareExchange(
+      expected: 0,
+      desired: _gettid(),
+      successOrdering: .acquiring,
+      failureOrdering: .relaxed
+    ).exchanged
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func unlock() {
+    // TODO: Is it worth caching this value in TLS?
+    let selfId = _gettid()
+    // Unconditionally update the storage to be unlocked. At this point, the
+    // lock is acquirable.
+    let oldValue = storage.exchange(0, ordering: .releasing)
```

Comment:
I don't think this is right. If `FUTEX_WAITERS` is set, we aren't supposed to unlock like this from user space — instead, we have to use `FUTEX_UNLOCK_PI`. At least, that's my understanding from the futex documentation.

We should only set it to zero where its current value is `selfId`, surely?

Reply:
You're right, we need to cmpxchg from gettid -> 0 and if there are waiters call futex unlock. I've gone ahead and updated.